### PR TITLE
Extract dask and spark test into distributed test.

### PR DESCRIPTION
**Pull request by trivialfis**
- Move test files.
- Run spark and dask separately to prevent conflicts.
- Gather common code into testing module.

### CODE REVIEW

**Comment by trivialfis** [comment]
Close https://github.com/dmlc/xgboost/issues/8134 .

File: tests/test_distributed/test_federated/test_federated.py [hunk 1]
```diff
@@ -80,7 +81,7 @@ def run_test(with_ssl: bool = True, with_gpu: bool = False) -> None:
 
 
 if __name__ == '__main__':
-    run_test(with_ssl=True, with_gpu=False)
-    run_test(with_ssl=False, with_gpu=False)
-    run_test(with_ssl=True, with_gpu=True)
-    run_test(with_ssl=False, with_gpu=True)
+    run_federated(with_ssl=True, with_gpu=False)
```
Comment (trivialfis) [inline tests/test_distributed/test_federated/test_federated.py:84]: We should convert these into pytest and run them on CI.

**Comment by trivialfis** [comment]
Should I put the tests as part of XGBoost module?
Pros:
- Users can run the tests, especially helpful if they are troubleshooting distributed environment.
- Distributors can run the tests. This way packaging systems like conda doesn't need to download the git repository or invent their own tests.
- Easier imports when writing tests.

Cons:
- Size of the Python package will grow.
- We need some heavy changes to annotate the tests with Python type hint and make the linter happy.

Some other libraries like dask/distributed, cudf, and cuml are integrating the tests into their Python module. For more implications, see https://docs.pytest.org/en/7.1.x/explanation/goodpractices.html#choosing-a-test-layout-import-rules .

**Comment by trivialfis** [comment]
Let's keep the existing layout for now. If the need ever comes up we can make the change then, git can recognize moving files.

**Review by hcho3 — APPROVED** [review]

**Pull request merged**